File: MultipleLinearRegression1.R
#MULTIPLE LINEAR REGRESSION
dataset = read.csv('50_Startups.csv')
dataset$State = factor(dataset$State, levels =c('New York', 'California', 'Florida'),
                         labels = c(1, 2 ,3))
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
train_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

#Fitting and predicting(Simple linear regression with all variables)
regressor = lm(formula = Profit ~ .,
               data = train_set)
y_pred = predict(regressor, newdata = test_set)

#Fitting and predicting(Multiple linear regression with 1 variable)
regressor_multi = lm(formula = Profit ~ R.D.Spend,
               data = train_set)
y_pred_multi = predict(regressor, newdata = test_set)

#Backtracking
regressor_backtrack = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
                     data = train_set)
summary(regressor_backtrack)

backwardElimination <- function(x, sl) {
  numVars = length(x)
  for (i in c(1:numVars)){
    regressor = lm(formula = Profit ~ ., data = x)
    maxVar = max(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"])
    if (maxVar > sl){
      j = which(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"] == maxVar)
      x = x[, -j]
    }
    numVars = numVars - 1
  }
  return(summary(regressor))
}

SL = 0.05
dataset = dataset[, c(1,2,3,4,5)]
backwardElimination(train_set, SL)



#Visualising

File: MultipleLinearRegression1.py
#MULTIPLE LINEAR REGRESSION

#data preprocessing
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
ct = ColumnTransformer([('encoder', OneHotEncoder(), [3])], remainder='passthrough')
X = np.array(ct.fit_transform(X), dtype=np.float)

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 1/3, random_state = 0)

#Fitting
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

#Predicting
y_pred = regressor.predict(X_test)

#Backward Elimination
import statsmodels.formula.api as sm
X = np.append(arr =  np.ones((50,1)).astype(int),values=X,axis=1)
X_optimal = X[:, [0,1,2,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_optimal).fit()
print(regressor_OLS.summary())

#automated Backwad elimination
def backwardElimination(x, sl):
    numVars = len(x[0])
    for i in range(0, numVars):
        regressor_OLS = sm.OLS(y, x).fit()
        maxVar = max(regressor_OLS.pvalues).astype(float)
        if maxVar > sl:
            for j in range(0, numVars - i):
                if (regressor_OLS.pvalues[j].astype(float) == maxVar):
                    x = np.delete(x, j, 1)
    regressor_OLS.summary()
    return x
 
SL = 0.05
X_opt = X[:, [0, 1, 2, 3, 4, 5]]
X_Modeled = backwardElimination(X_opt, SL)

regressor_OLS = sm.OLS(endog = y, exog = X_Modeled).fit()
y_pred = regressor.predict(X_test)

#Visualising
plt.scatter
